File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def bfs(maps, visited):
    q = deque([(0, 0, 0)])
    visited[0][0] = True

    while q:
        x, y, dist = q.popleft()
        dx = [0, 0, -1, 1]
        dy = [-1, 1, 0, 0]
        
        # 목표 도달 시
        if x == len(maps) - 1 and y == len(maps[0]) - 1:
            return dist + 1
        
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < len(maps) and 0 <= ny < len(maps[0]) and not visited[nx][ny] and maps[nx][ny] != 0:
                visited[nx][ny] = True
                q.append((nx, ny, dist + 1))

    return -1  # 경로가 없을 경우 -1 반환


def solution(maps):
    visited = [[False] * len(maps[0]) for _ in range(len(maps))]
    answer = bfs(maps, visited)
    return answer  # 경로가 없으면 -1이 반환됨

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include <vector>
#include <tuple>
#include <deque>

using namespace std;

int bfs(vector<vector<int>>& maps) {
    int n = maps.size();
    int m = maps[0].size();
    
    deque<tuple<int, int, int>> dq;
    vector<vector<bool>> visited(n, vector<bool>(m, false));
    
    visited[0][0] = true;
    dq.push_back({0, 0, 1}); 

    int dx[] = {0, 0, -1, 1};
    int dy[] = {-1, 1, 0, 0};
    
    while (!dq.empty()) {
        tuple<int, int, int> curr = dq.front();
        dq.pop_front();
        
        int x = get<0>(curr);
        int y = get<1>(curr);
        int dist = get<2>(curr);

        if (x == n - 1 && y == m - 1) {
            return dist;
        }
        
        for (int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];
            
            if (nx >= 0 && nx < n && ny >= 0 && ny < m && !visited[nx][ny] && maps[nx][ny] == 1) {
                visited[nx][ny] = true;
                dq.push_back({nx, ny, dist + 1});
            }
        }
    }
    return -1;
}

int solution(vector<vector<int>> maps) {
    return bfs(maps);
}